### PR TITLE
CLOUDP-276381: adding formatting to SDK

### PR DESCRIPTION
**Pull request by wtrocki**
## Description

Adding missing SDK error formatting introduced recently by us in backend.

## Review details

1. Review example file 
2. Tests should be passing (example is an test)
3. Review template

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code

### CODE REVIEW

File: admin/client.go [hunk 1]
```diff
@@ -691,9 +691,15 @@ func (e *GenericOpenAPIError) SetError(errorString string) {
...
 }
 
-// format error message using title and detail when model implements Error
-func formatErrorMessage(status, path, method string, v ApiError) string {
-	return fmt.Sprintf("%v %v: HTTP %v (Error code: %q) Detail: %v Reason: %v. Params: %v",
+// FormatErrorMessageWithDetails: format error message using error struct fields: Should be only used for testing
+func FormatErrorMessageWithDetails(status, path, method string, v ApiError) string {
+	detailString := ""
+	if v.BadRequestDetail != nil {
+		badRequestDetail, _ := json.Marshal(v.GetBadRequestDetail())
```
Comment (lantoli) [inline admin/client.go:698]: do we want to check for errors in case JSON is incorrect?

File: admin/client.go [hunk 2]
```diff
@@ -691,9 +691,15 @@ func (e *GenericOpenAPIError) SetError(errorString string) {
 	e.error = errorString
 }
 
-// format error message using title and detail when model implements Error
-func formatErrorMessage(status, path, method string, v ApiError) string {
-	return fmt.Sprintf("%v %v: HTTP %v (Error code: %q) Detail: %v Reason: %v. Params: %v",
+// FormatErrorMessageWithDetails: format error message using error struct fields: Should be only used for testing
```
Comment (lantoli) [inline admin/client.go:694]:
```suggestion
// FormatErrorMessageWithDetails formats error message using error struct fields: Should be only used for testing
```

more Go idiomatic for func comments

File: admin/client.go [hunk 3]
```diff
@@ -691,9 +691,15 @@ func (e *GenericOpenAPIError) SetError(errorString string) {
...
+	detailString := ""
+	if v.BadRequestDetail != nil {
+		badRequestDetail, _ := json.Marshal(v.GetBadRequestDetail())
+		detailString = string(badRequestDetail)
+	}
+
+	return fmt.Sprintf("%v %v: HTTP %v (Error code: %q) Detail: %v Reason: %v. Params: %v, BadRequestDetail: %v",
 		method, path, status, v.GetErrorCode(),
-		v.GetDetail(), v.GetReason(), v.GetParameters())
+		v.GetDetail(), v.GetReason(), v.GetParameters(), detailString)
```
Comment (lantoli) [inline admin/client.go:704]: knit: we have "details" name for two things, don't know if we could change one of them

File: tools/config/go-templates/client.mustache [hunk 4]
```diff
@@ -652,9 +652,16 @@ func (e *GenericOpenAPIError) SetError(errorString string) {
 	e.error = errorString;
 }
 
-// format error message using title and detail when model implements Error
-func formatErrorMessage(status, path, method string, v ApiError) string {
-	return fmt.Sprintf("%v %v: HTTP %v (Error code: %q) Detail: %v Reason: %v. Params: %v",
-			method, path, status, v.GetErrorCode(),
-	 		v.GetDetail(), v.GetReason(), v.GetParameters())
+// FormatErrorMessageWithDetails: format error message using error struct fields: Should be only used for testing
+func FormatErrorMessageWithDetails(status, path, method string, v ApiError) string {
```
Comment (lantoli) [inline tools/config/go-templates/client.mustache:656]: i guess my suggestion for func comment should be changed here in the template

File: admin/client.go [hunk 5]
```diff
@@ -691,9 +691,15 @@ func (e *GenericOpenAPIError) SetError(errorString string) {
...
 
-// format error message using title and detail when model implements Error
-func formatErrorMessage(status, path, method string, v ApiError) string {
-	return fmt.Sprintf("%v %v: HTTP %v (Error code: %q) Detail: %v Reason: %v. Params: %v",
+// FormatErrorMessageWithDetails: format error message using error struct fields: Should be only used for testing
+func FormatErrorMessageWithDetails(status, path, method string, v ApiError) string {
+	detailString := ""
+	if v.BadRequestDetail != nil {
+		badRequestDetail, _ := json.Marshal(v.GetBadRequestDetail())
+		detailString = string(badRequestDetail)
```
Comment (lantoli) [inline admin/client.go:699]:
```suggestion
		if badRequestDetail, err := json.Marshal(v.GetBadRequestDetail()); err != nil {
		   detailString = string(badRequestDetail)
		}
```

shouldn't we check for errors converting JSON?

**Review by lantoli — COMMENTED** [review]
LGTM, just a few minor comments, note that linter is failing

**Review by lantoli — APPROVED** [review]

File: admin/client.go [hunk 6]
```diff
@@ -691,9 +691,15 @@ func (e *GenericOpenAPIError) SetError(errorString string) {
 	e.error = errorString
 }
 
-// format error message using title and detail when model implements Error
-func formatErrorMessage(status, path, method string, v ApiError) string {
-	return fmt.Sprintf("%v %v: HTTP %v (Error code: %q) Detail: %v Reason: %v. Params: %v",
+// FormatErrorMessageWithDetails: format error message using error struct fields: Should be only used for testing
+func FormatErrorMessageWithDetails(status, path, method string, v ApiError) string {
```
Comment (lantoli) [inline admin/client.go:695]: can't it be used in prod code, e.g. to show Terraform errors to the user?

File: admin/client.go (hunk 5 above)
Comment (wtrocki) [reply admin/client.go:699]: We do parse JSON first from backend response. Even if we had invalid JSON  at this point somehow by regression in future would you see us doing different handling here? TL:DR - SDK does have request logger that shows rawa data. I do not like to throw errors from error handling method

File: admin/client.go (hunk 6 above)
Comment (wtrocki) [reply admin/client.go:695]: https://github.com/mongodb/atlas-sdk-go/blob/main/docs/doc_2_error_handling.md 

With that context I do not think we need anyone to use that method directly. Exposing that method will give us ability to do better unit tests in cli/TF as we can autoformat error message (right now we have those values hardcoded which is maintenance burden every time test changes).

File: tools/config/go-templates/client.mustache (hunk 4 above)
Comment (wtrocki) [reply tools/config/go-templates/client.mustache:656]: Yes. I always accept comment in 1 place and regenerate locally. 
Normal approach for SDK.

File: admin/client.go (hunk 3 above)
Comment (wtrocki) [reply admin/client.go]: Done

File: admin/client.go (hunk 6 above)
Comment (lantoli) [reply admin/client.go:695]: ok

**Review by lantoli — APPROVED** [review]

**Comment by wtrocki** [comment]
Thank you for review @lantoli

**Pull request merged**